File: cpp/3243_Shortest_Distance_After_Road_Addition_Queries_I.cpp
class Solution {
public:
    vector<int> shortestDistanceAfterQueries(int n, vector<vector<int>>& queries) {
        vector<vector<int>> adj(n);
        for (int i = 0; i < n - 1; i++)
            adj[i].push_back(i + 1);

        vector<int> ans;
        
        auto bfs = [&](int start) 
        {
            vector<int> dist(n, -1);
            queue<int> q;
            q.push(start);
            dist[start] = 0;
            while (!q.empty()) 
            {
                int u = q.front(); q.pop();
                if (u == n - 1) return dist[u];
                for (int nxt : adj[u]) 
                {
                    if (dist[nxt] == -1) 
                    {
                        dist[nxt] = dist[u] + 1;
                        q.push(nxt);
                    }
                }
            }
            return dist[n - 1];
        };
        
        for (auto &q : queries) 
        {
            adj[q[0]].push_back(q[1]);
            ans.push_back(bfs(0));
        }
        
        return ans;
    }
};

File: py/3243_Shortest_Distance_After_Road_Addition_Queries_I.py
from collections import deque

class Solution(object):
    def shortestDistanceAfterQueries(self, n, queries):
        """
        :type n: int
        :type queries: List[List[int]]
        :rtype: List[int]
        """
        adj = [[] for _ in range(n)]

        for i in range(n - 1):
            adj[i].append(i + 1)

        def bfs():
            dist = [-1] * n
            q = deque([0])
            dist[0] = 0 

            while q:
                u = q.popleft()
                if u == n - 1:
                    return dist[u]

                for v in adj[u]:
                    if dist[v] == -1:
                        dist[v] = dist[u] + 1
                        q.append(v)

            return dist[n - 1]

        ans = []

        for u, v in queries:
            adj[u].append(v)
            ans.append(bfs())

        return ans
        
